src/mpi/coll/transports/generic/transport.c

```diff
@@ -0,0 +1,1243 @@
+
+                break;
+
+            default:
+                break;
+            }
+        }
+    }
+    /* set the tail of the issued vertex linked lsit */
+    sched->last_issued->next_issued = NULL;
+
+#ifdef MPIR_COLL_DEBUG
+    if (sched->num_completed == sched->total) {
+        MPL_DBG_MSG_FMT(MPIR_DBG_COLL,VERBOSE,(MPL_DBG_FDEST,"  --> GENERIC transport (test) complete:  sched->total=%d\n", sched->total));
+    }
+#endif
+
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIR_COLL_GENERIC_POKE);
+
+    /* return 1 if complete, else return 0 */
+    return sched->num_completed == sched->total;
+}
+
+/* Internal transport function to free any memory allocated for execution of this schedule */
+MPL_STATIC_INLINE_PREFIX void MPIR_COLL_GENERIC_free_buffers(MPIR_COLL_GENERIC_sched_t * sched)
+{
+    int i;
+    for (i = 0; i < sched->total; i++) {
+        /* free the temporary memory allocated by recv_reduce call */
+        if (sched->vtcs[i].kind == MPIR_COLL_GENERIC_KIND_RECV_REDUCE) {
+            MPIR_COLL_GENERIC_free_mem(sched->vtcs[i].nbargs.recv_reduce.inbuf);
+            sched->vtcs[i].nbargs.recv_reduce.inbuf = NULL;
+        }
+
+        if (sched->vtcs[i].kind == MPIR_COLL_GENERIC_KIND_FREE_MEM) {
+            MPIR_COLL_GENERIC_free_mem(sched->vtcs[i].nbargs.free_mem.ptr);
+            sched->vtcs[i].nbargs.free_mem.ptr = NULL;
+        }
+
+        if(sched->vtcs[i].kind == MPIR_COLL_GENERIC_KIND_MULTICAST){
+            MPIR_COLL_GENERIC_free_mem(sched->vtcs[i].nbargs.multicast.mpir_req);
+            MPIR_COLL_GENERIC_free_mem(sched->vtcs[i].nbargs.multicast.destinations);
+        }
+    }
+    /* free temporary buffers allocated using MPIR_COLL_GENERIC_allocate_buffer call */
+    for (i = 0; i < sched->buf_array.used; i++) {
+        MPIR_COLL_GENERIC_free_mem(sched->buf_array.array[i]);
+    }
+    if(sched->buf_array.size!=0)
+        MPIR_COLL_GENERIC_free_mem(sched->buf_array.array);
+
+    /* free each vtx and then the list of vtcs */
+    for (i = 0; i < sched->total; i++) {
+        /* up to sched->total because we init vertices only when we need them */
+        MPIR_COLL_GENERIC_free_vtx(&sched->vtcs[i]);
+    }
+    /* free the list of vertices */
+    MPIR_COLL_GENERIC_free_mem(sched->vtcs);
+}
+
+/* Internal transport function to destroy all memory associated with a schedule */
+MPL_STATIC_INLINE_PREFIX void MPIR_COLL_GENERIC_sched_destroy_fn(MPIR_COLL_GENERIC_sched_t * sched)
+{
+    MPL_DBG_MSG_FMT(MPIR_DBG_COLL,VERBOSE,(MPL_DBG_FDEST,"In MPIR_COLL_GENERIC_sched_destroy_fn for schedule: %p\n", sched));
+    MPIR_COLL_GENERIC_free_buffers(sched);
+    MPL_free(sched);
+}
+
+/* Transport function to initialize a transport communicator */
+int MPIR_COLL_GENERIC_comm_init(MPIR_COLL_GENERIC_comm_t * comm, void *base)
+{
+    MPIR_Comm *mpi_comm = container_of(base, MPIR_Comm, coll);
+    comm->mpid_comm = mpi_comm;
+    comm->sched_db = NULL;
+    return 0;
```

Comment:
Why the `return 0;` here? Should it be `return MPI_SUCCESS;`?

Reply:
Done